### PR TITLE
[RISCV][GISel] Add support for G_FCMP with F and D extensions.

### PR DESCRIPTION
**Pull request by topperc**
We only have instructions for OEQ, OLT, and OGT. We need to convert other comparison codes into those.

I think we'll likely want to split this up in the future to support optimizations. Maybe do some of it in the legalizer or in a new post legalizer lowering pass. So this patch is just enough to get something working without adding 11 additional patterns to tablegen for each type.

### CODE REVIEW

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 1]
```diff
@@ -674,6 +689,123 @@ bool RISCVInstructionSelector::selectSelect(MachineInstr &MI,
...
+  case CmpInst::FCMP_OLE:
+    return Size == 32 ? RISCV::FLE_S : RISCV::FLE_D;
+  case CmpInst::FCMP_OEQ:
+    return Size == 32 ? RISCV::FEQ_S : RISCV::FEQ_D;
+  }
+}
+
+// Try legalizing an FCMP by swapping or inverting the predicate to one that
+// is supported.
+static bool legalizeFCmpPredicate(Register &LHS, Register &RHS,
```
Comment (arsenm) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:709]: This kind of stuff should go in LegalizerHelper. We should be able to add the cond code to the legalizer queries for G_ICMP/G_FCMP and replicate the DAG's setCondCodeAction

Comment (tschuett) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:709]: It should be fine to claim support for all condition codes in the legalizer. Do target independent combines and later down the pass pipeline, they can do RISCV-V specific optimizations.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 2]
```diff
@@ -67,6 +67,9 @@ class RISCVInstructionSelector : public InstructionSelector {
   bool selectSelect(MachineInstr &MI, MachineIRBuilder &MIB,
                     MachineRegisterInfo &MRI) const;
 
+  bool emitFPCompare(CmpInst::Predicate Pred, Register DstReg, Register LHS,
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:70]: Should this function be renamed to `selectFPCompare`? It looks like we are using it to select a G_FCMP.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 3]
```diff
@@ -366,6 +369,18 @@ bool RISCVInstructionSelector::select(MachineInstr &MI) {
...
+  case TargetOpcode::G_FCMP: {
+    CmpInst::Predicate Pred =
+        static_cast<CmpInst::Predicate>(MI.getOperand(1).getPredicate());
+
+    if (!emitFPCompare(Pred, MI.getOperand(0).getReg(),
+                       MI.getOperand(2).getReg(), MI.getOperand(3).getReg(),
+                       MIB))
+      return false;
+
+    MI.eraseFromParent();
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:381]: Should we move the `eraseFromParent` into `emitFPCompare` like the other select functions do?

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 4]
```diff
@@ -674,6 +689,123 @@ bool RISCVInstructionSelector::selectSelect(MachineInstr &MI,
...
+}
+
+// Try legalizing an FCMP by swapping or inverting the predicate to one that
+// is supported.
+static bool legalizeFCmpPredicate(Register &LHS, Register &RHS,
+                                  CmpInst::Predicate &Pred, bool &NeedInvert) {
+  auto isLegalFCmpPredicate = [](CmpInst::Predicate Pred) {
+    return Pred == CmpInst::FCMP_OLT || Pred == CmpInst::FCMP_OLE ||
+           Pred == CmpInst::FCMP_OEQ;
+  };
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:714]: Do you need to check the case where `Pred` is already legal: `isLegalFCmpPredicate(Pred)`?

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 5]
```diff
@@ -674,6 +689,123 @@ bool RISCVInstructionSelector::selectSelect(MachineInstr &MI,
...
+  assert((Size == 32 || Size == 64) && "Unexpected size");
+
+  Register TmpReg = DstReg;
+
+  bool NeedInvert = false;
+  // First try swapping operands or inverting.
+  if (legalizeFCmpPredicate(LHS, RHS, Pred, NeedInvert)) {
+    if (NeedInvert)
+      TmpReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+    auto Cmp = MIB.buildInstr(getFCmpOpcode(Pred, Size), {TmpReg}, {LHS, RHS});
```
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:758]: `MIB.buildFCmp(getFCmpOpcode(Pred, Size), TmpReg, LHS, RHS)` and below?

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:758]: Doesn't buildFcmp create a G_FCMP?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:758]: You are right. My bad.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp (hunk 4 above)
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:703]: I see that `olt` is getting selected correctly, but I don't see it getting handled by any of the code in this patch. How is that code getting selected? `selectImpl`?

Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:703]: Yes. tablegen has patterns for oeq, olt, and ole so selectImpl gets it.

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:703]: Got it. If that’s the case, then it’s up to you if you’d like to have the assert here. I made my initial comment thinking we had to handle it with custom code.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 6]
```diff
@@ -674,6 +678,130 @@ bool RISCVInstructionSelector::selectSelect(MachineInstr &MI,
...
+// FIXME: Maybe we should expand this earlier.
+bool RISCVInstructionSelector::selectFPCompare(MachineInstr &MI,
+                                               MachineIRBuilder &MIB,
+                                               MachineRegisterInfo &MRI) const {
+  CmpInst::Predicate Pred =
+      static_cast<CmpInst::Predicate>(MI.getOperand(1).getPredicate());
+
+  Register DstReg = MI.getOperand(0).getReg();
+  Register LHS = MI.getOperand(2).getReg();
+  Register RHS = MI.getOperand(3).getReg();
```
Comment (aemerson) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:741]: We have the `GICmp/GFCmp` wrapper classes that you can use. They provide things like `getCond()`, `getLHSReg()`. To use it just cast the `MachineInstr`: `auto &Cmp = cast<GFCmp>(MI);`

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 7]
```diff
@@ -836,6 +840,130 @@ bool RISCVInstructionSelector::selectSelect(MachineInstr &MI,
...
+    if (!Cmp2.constrainAllUses(TII, TRI, RBI))
+      return false;
+    if (NeedInvert)
+      TmpReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+    auto And =
+        MIB.buildInstr(RISCV::AND, {TmpReg}, {Cmp1.getReg(0), Cmp2.getReg(0)});
+    if (!And.constrainAllUses(TII, TRI, RBI))
+      return false;
+  } else
+    llvm_unreachable("Unhandled predicate");
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:954]: style: add braces here to keep it uniform with prior if blocks

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp [hunk 8]
```diff
@@ -836,6 +840,130 @@ bool RISCVInstructionSelector::selectSelect(MachineInstr &MI,
...
+  bool NeedInvert = false;
+  // First try swapping operands or inverting.
+  if (legalizeFCmpPredicate(LHS, RHS, Pred, NeedInvert)) {
+    if (NeedInvert)
+      TmpReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+    auto Cmp = MIB.buildInstr(getFCmpOpcode(Pred, Size), {TmpReg}, {LHS, RHS});
+    if (!Cmp.constrainAllUses(TII, TRI, RBI))
+      return false;
+  } else if (Pred == CmpInst::FCMP_ONE || Pred == CmpInst::FCMP_UEQ) {
+    // fcmp ogt LHS, RHS => (OR (FLT LHS, RHS), (FLT RHS, LHS))
```
Comment (mshockwave) [inline llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:919]: `fcmp uno`?

**Review by mshockwave — APPROVED** [review]
LGTM w/ minor comments.

I'm fine with keeping `legalizeFCmpPredicate` for now, as long as it will eventually be replaced by LegalizeHelper.

File: llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp (hunk 8 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp:919]: `fcmp one`?

**Pull request merged**